magicgui/_magicgui.py

```diff
@@ -144,36 +187,176 @@ def magicgui(
-        )
-        func_gui.__wrapped__ = func
-        return func_gui
+    Examples
+    --------
+    >>> def func(x: int, y: str):
+    ...     pass
+    ...
+    >>> factory = MagicFactory(function=func, labels=False)
+    >>> # factory accepts all the same arguments as magicgui()
+    >>> widget1 = factory(call_button=True)
+    >>> # can also override magic_kwargs that were provided when creating the factory
+    >>> widget2 = factory(auto_call=True, labels=True)
+    """
 
-    if function is None:
-        return inner_func
-    else:
-        return inner_func(function)
+    def __new__(cls, function, *args, magic_class=FunctionGui, **keywords):
+        """Create new MagicFactory."""
+        if not function:
+            raise TypeError(
+                "MagicFactory missing required positional argument 'function'"
+            )
+        if isinstance(function, FunctionType):
+            if "<locals>" in function.__qualname__:
+                freevars = function.__code__.co_freevars
+                if function.__name__ in freevars:
+                    warn(
+                        "Self-reference detected in MagicFactory function created "
+                        "in a local scope. FunctionGui references will not work."
+                    )
+        # we want function first for the repr
+        keywords = {"function": function, **keywords}
+        return super().__new__(cls, magic_class, *args, **keywords)  # type: ignore
+
+    def __repr__(self) -> str:
+        """Return string repr."""
+        params = inspect.signature(magicgui).parameters
+        args = [
+            f"{k}={v!r}"
+            for (k, v) in self.keywords.items()
+            if v not in (params[k].default, {})
+        ]
+        return f"MagicFactory({', '.join(args)})"
+
+    def __call__(self, *args, **kwargs) -> _T:
+        """Call the wrapped _magicgui and return a FunctionGui."""
+        if args:
+            raise ValueError("MagicFactory instance only accept keyword arguments")
+        params = inspect.signature(magicgui).parameters
+        prm_options = self.keywords.pop("param_options", {})
+        prm_options.update({k: kwargs.pop(k) for k in list(kwargs) if k not in params})
+        return self.func(param_options=prm_options, **{**self.keywords, **kwargs})
+
+    def __getattr__(self, name) -> Any:
+        """Allow accessing FunctionGui attributes without mypy error."""
+        pass  # pragma: no cover
+
+    @property
+    def __name__(self) -> str:
+        """Pass function name."""
+        return getattr(self.keywords.get("function"), "__name__", "FunctionGui")
+
+
+# Overloads for magic_factory decorator.  See implementation below
+# TODO: figure out how to get these in a stub file.
+# My first attempts to put them in ``_magicgui.pyi`` broke my type hints in VSCode
+# fmt: off
+@overload  # noqa: E302
+def magic_factory(  # noqa
+    function: Callable[..., _R],
+    *,
+    layout: str = "horizontal",
```

Comment:
Have we not settled on vertical default yet?

Reply:
not yet, but I could be convinced.

Reply:
#121